File: Data Structures & Algorithms/two-integer-sum-ii/submission-3.java
class Solution {
    public int[] twoSum(int[] numbers, int target) {
        int l = 0;
        int r = numbers.length - 1;
        while (l < r) {
            // if (numbers[l] > target) {
            //     // System.out.println(l + " : " + numbers[l]);
            //     return null;
            // }
            // else if (numbers[r] < target) {
            //     System.out.println(r + " : " + numbers[r]);
            //     return null;
            // }
            // else {
                if (target - numbers[l] == numbers[r]) {
                    return new int[]{l+1, r+1};
                }
                else if (target - numbers[l] < numbers[r]) {
                    // System.out.println("r = " + r + "--");
                    r--;
                }
                else {
                    // System.out.println("l = " + l + "++");
                    l++;
                }
            // }
        }
        // System.out.println("none");
        return null;
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-3.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res_set = set()
        res = []
        nums = sorted(nums)
        self.helper(nums, target, 0, res_set, [], 0)

        for element in res_set:
            res.append(list(element))
        return res
        
    def helper(self, nums: List[int], target: int, curr: int, res_set: set, curr_progress: list, index: int) -> None:
        for i in range(index, len(nums)):
            num = nums[i]
            curr += num
            curr_progress.append(num)
            if curr == target:
                res_set.add(tuple(sorted(curr_progress)))
            elif curr > target:
                curr_progress.pop(-1)
                return
            else:
                self.helper(nums, target, curr, res_set, curr_progress, i)
            curr_progress.pop(-1)
            curr -= num
        return